Tugas 2 Prak ML A1

Muhammad Haikal Ali| 24060121130064 | ML A1

## No 1

Dataset : Bunga Iris

Link: http://archive.ics.uci.edu/dataset/53/iris

In [2]:
import pandas as pd

In [3]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'species']

dataset = pd.read_csv(url, names=names)

In [4]:
# Import library yang dibutuhkan
from sklearn import model_selection

In [5]:
# Membagi dataset validasi
array = dataset.values
X = array[:,0:4]
Y = array[:,4]
validation_size = 0.20

seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

Penjelasan :

array = dataset.values: Ini mengkonversi dataset ke dalam bentuk array. dataset di sini mungkin objek yang berisi data, dan values mungkin digunakan untuk mengakses nilai-nilai sebenarnya dari dataset tersebut.

X = array[:,0:4]: Ini mengambil kolom 0 hingga 3 dari array sebagai fitur (input) model. Dalam konteks machine learning, ini mungkin berarti bahwa kolom 0 hingga 3 adalah variabel independen atau fitur yang akan digunakan untuk melakukan prediksi.

Y = array[:,4]: Ini mengambil kolom ke-4 dari array sebagai target (output) model. Dalam konteks machine learning, ini mungkin berarti bahwa kolom ke-4 berisi nilai yang ingin kita prediksi.

validation_size = 0.20: Variabel ini menentukan ukuran data validasi sebagai proporsi dari keseluruhan dataset. Dalam hal ini, 20% dari data akan digunakan sebagai data validasi.

seed = 7: Ini adalah nilai seed yang digunakan dalam fungsi pengacakan (random) untuk memastikan hasil yang konsisten. Jika Anda menggunakan nilai seed yang sama, Anda akan mendapatkan pembagian data yang sama setiap kali Anda menjalankan kode ini.

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed): Ini menggunakan fungsi train_test_split dari pustaka model_selection (mungkin dari scikit-learn) untuk membagi dataset menjadi dua bagian: data latih (X_train dan Y_train) dan data validasi (X_validation dan Y_validation). Proporsi data validasi ditentukan oleh test_size, dan nilai seed digunakan untuk pengacakan agar pembagian data dapat direproduksi.

Jadi, secara keseluruhan, kode ini bertujuan untuk mempersiapkan data untuk pelatihan dan validasi model machine learning dengan membagi dataset menjadi fitur dan target, serta membagi data menjadi dua bagian: data latih dan data validasi.

In [6]:
# Melihat dataset validasi yang sudah dibagi
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (120, 4)
X_validation : (30, 4)
Y_train : (120,)
Y_validation : (30,)


In [7]:
# Mendefinisikan opsi tes dan metrik evaluasi
seed = 7

scoring = 'accuracy'

In [8]:
# Import library berisi algoritma klasifikasi yang akan digunakan
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import SGDClassifier

models = []

models.append(('RF', RandomForestClassifier()))
models.append(('DT', DecisionTreeClassifier()))
models.append(('SGD', SGDClassifier()))

In [9]:
# Melakukan evaluasi pada setiap model

results = []

names = []

for name, model in models:

  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle = True)

  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)

  results.append(cv_results)

  names.append(name)

  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

RF: 0.966667 (0.040825)
DT: 0.950000 (0.076376)
SGD: 0.933333 (0.062361)


Jika sudah memiliki hasil evaluasi dari ketiga model diatas untuk memilih model terbaik dilakukan dari perbandingan satu sama lainnya dan dipilih yang paling akurat. Dari eksekusi script diatas, kita mendapatkan hasil mentah sebagai berikut:

  RF: 0.966667 (0.040825)

  DT: 0.950000 (0.076376)
  
  SGD: 0.933333 (0.062361)

  Dari hasil di atas, kita dapat melihat bahwa RandomForest Classifier memiliki nilai akurasi perkiraan terbesar. Setelah mengetahui model yang paling akurat yaitu RandomForestClassifier (RF), selanjutnya kita dapat mencoba melakukan pengujian tentang keakuratan model RandomForest Classifier terhadap data yang ada.

In [10]:
# Import library yang dibutuhkan
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [12]:
# Membuat prediksi pada dataset validasi

rf = RandomForestClassifier()

rf.fit(X_train, Y_train)

predictions = rf.predict(X_validation)

print(accuracy_score(Y_validation, predictions))

print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.8666666666666667
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



Kesimpulan

Dari hasil pengujian diatas, untuk dataset Bunga Iris didapatkan nilai accuracy pada data testing adalah 0.8666666666666667

## No 2

Dataset : Auto MPG

Link : "https://archive.ics.uci.edu/static/public/9/data.csv"

In [13]:
# import library pandas yang digunakan untuk mengelola dataset
import pandas as pd

In [14]:
# Memuat data set

# url disini merupakan parameter untuk menyimpan link yang merujuk datasetnya
url = "https://archive.ics.uci.edu/static/public/9/data.csv"

# names disini disini merupakan parameter untuk menyimpan array nama kolomnya nanti
names = ['car_name', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'mpg']

# param dataset disini digunakan untuk menyimpan data dari dataset yang sudah ada di param url yang dibaca dengan fungsi .read_csv yang dimiliki library pandas
DataFrame = pd.read_csv(url)

#mengubah nama kolom sesuai parameter names
DataFrame.rename(columns={'car_name': names[0], 'cylinders': names[1], 'displacement': names[2], 'horsepower': names[3], 'weight': names[4], 'acceleration': names[5], 'model_year': names[6], 'origin': names[7], 'mpg': names[8]}, inplace=True)

In [15]:
DataFrame.head()

car_name  cylinders  displacement  horsepower  weight  \
0  chevrolet,chevelle,malibu          8         307.0       130.0    3504   
1          buick,skylark,320          8         350.0       165.0    3693   
2         plymouth,satellite          8         318.0       150.0    3436   
3              amc,rebel,sst          8         304.0       150.0    3433   
4                ford,torino          8         302.0       140.0    3449   

   acceleration  model_year  origin   mpg  
0          12.0          70       1  18.0  
1          11.5          70       1  15.0  
2          11.0          70       1  18.0  
3          12.0          70       1  16.0  
4          10.5          70       1  17.0

In [16]:
from sklearn import model_selection

In [17]:
DataFrame.info()

array = DataFrame.values
X = array[:,0:8]
Y = array[:,8]
validation_size = 0.20
seed = 7
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   car_name      398 non-null    object 
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   mpg           398 non-null    float64
dtypes: float64(4), int64(4), object(1)
memory usage: 28.1+ KB


In [18]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (318, 8)
X_validation : (80, 8)
Y_train : (318,)
Y_validation : (80,)


In [19]:
# Test options and evaluation metric
seed = 7
scoring = 'accuracy'

In [20]:
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB


models = []
models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))

In [21]:
results = []

names = []

for name, model in models:

  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle = True)

  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)#, error_score='raise')

  results.append(cv_results)

  names.append(name)

  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

ValueError: ignored

MAAF KAK INI GA NEMU CARA MEMBENARKAN ERROR NYA UNTUK SAAT INI NANTI KETIKA SUDAH BENAR MINTA IJIN UNTUK MENGUMPULKAN ULANG

TERIMA KASIH KA